### DISCUSSION TITLE
Server extension - not able to use Tokens provided by jupyterlab packages

### DISCUSSION THREAD

**Discussion by tomjakubowski** [jupyterlite/jupyterlite#1206] [Q&A] [answered]
Hello! I am working on a Jupyterlite 0.1.3 server extension. I followed the how-to guide [Create a server extension](https://jupyterlite.readthedocs.io/en/stable/howto/extensions/server.html).

My extension is apparently unable to resolve dependencies in its `requires` sections. When the extension is activated, it fails, saying:

> index.ts:518 Plugin '@prospectiveco/jupyterlite-nbconvert:plugin' failed to activate.
> Error: No provider for: @jupyterlab/translation:ITranslator.
 
This token is provided by @jupyterlab/translation, which appears to be loaded after my extension is. I suspect that my package is using a different Token than the one providing it, but I'm having trouble finding the smoking gun for that. I think I've set up the `jupyterlab/sharedPackages` section correctly for this, but I'm unsure.

I've also got my eyebrow raised at the `yarn.lock` file which includes duplicate and/or 3.6.x versions of several jupyterlab/lumino dependencies.

Repo is here: https://github.com/tomjakubowski/jupyterlite-extension-broken

And I will include some sections from `package.json` also:

```
  "dependencies": {
    "@jupyterlab/translation": "~3.5.3",
    "@jupyterlite/server": "~0.1.3",
    "@lumino/widgets": "^1.37.2"
  },
  "devDependencies": {
    "@babel/core": "^7.11.6",
    "@babel/preset-env": "^7.12.0",
    "@jupyterlab/builder": "~3.5.3",
    "@typescript-eslint/eslint-plugin": "^4.8.1",
    "@typescript-eslint/parser": "^4.8.1",
    "eslint": "^7.14.0",
    "eslint-config-prettier": "^6.15.0",
    "eslint-plugin-prettier": "^3.1.4",
    "npm-run-all": "^4.1.5",
    "prettier": "^2.1.1",
    "rimraf": "^3.0.2",
    "typescript": "4.9.5"
  },
  "jupyterlab": {
    "extension": true,
    "outputDir": "jupyterlite_nbconvert/labextension",
    "sharedPackages": {
      "@jupyterlab/translation": {
        "bundled": false,
        "singleton": true
      }
    }
  }
```

**Comment by tomjakubowski** [comment]
Following the same pattern for @jupyterlite dependencies (for example, `@jupyterlite/translation`), the extension is able to correctly load the tokens, as long it's been configured in `jupyterlab/sharedPackages`.

So it seems this is related specifically to tokens/packages/plugins which ship with Jupyterlab, such as `@jupyterlab/translation`'s ITranslator`.

**Reply by tomjakubowski** [reply]
Ah, I wonder if this is because it's a Lite server extension trying to require a dependency from a Lab front-end extension -- presumably the Lite server boots before the Lab front-end does.

**Reply by jtpio** [reply, marked as answer]
Most likely yes. Currently 2 apps are created on the frontend:

- a `JupyterLab` / `NotebookApp` / `SingleWidgetApp` for the regular frontend
- a `JupyterLiteServer` for lite server extensions

https://github.com/jupyterlite/jupyterlite/blob/2ecd4a117c5a150cbbf2859b3226f14ccbf9a0b4/app/index.template.js#L176-L189

But the regular JupyterLab extensions are registered on the main `JupyterLab` app, not `JupyterLiteServer`.

Wondering if in this case you could maybe try to provide `@jupyterlab/translation:ITranslator`, but from a lite server extension?

**Reply by tomjakubowski** [reply]
Thanks, the context is helpful! I believe our use case is served by writing two extensions:

1. a front-end extension which can access ITranslator, IMainMenu, ICommandPalette, etc.
2. a server extension which implements routes accessed by the front-end extension

Currently we've got this set up as two separate extensions (with their own python + node packaging).

Is it possible to combine the front-end and server extensions into one package? I know that we can export multiple server extensions like:

```
const alpha: JupyterLiteServerPlugin<void> = {...};
const beta: JupyterLiteServerPlugin<void> = {...};
const plugins = [alpha, beta];
```

But at least starting from the cookiecutter template, it doesn't seem like exporting an array with both `JupyterliteServerPlugin` and `JupyterliteFrontEnd` types in it does what I want - Jupyterlite tries loading both of them as server extensions.
